src/components/SeqVizParamsModal.vue

```diff
@@ -1,118 +1,271 @@
-                        type="button"
-                        class="btn btn-primary">
-                        Save changes
-                    </button>
-                    <button
-                        v-if="loadingInstance"
-                        v-on:click="$emit('submitInstance')"
-                        type="button"
-                        class="btn btn-primary">
-                        Add this sequence
-                    </button>
-                    <button
-                        v-on:click="$emit('closeModal')"
-                        type="button"
-                        class="btn btn-secondary"
-                        data-dismiss="modal">
-                        Close
-                    </button>
-                </div>
+                        <small
+                            v-bind:id="name + '-help-text'"
+                            class="form-text text-muted">
+                            {{ param.description }}
+                        </small>
+                    </div>
+                    <small v-if="Object.values(params).some(v => v.required)">
+                        <span class="redtext">*</span>Parameters with an
+                        asterisk are required.
+                    </small>
+                </form>
+            </div>
+            <div class="modal-footer">
+                <button
+                    v-if="!loadingInstance"
+                    v-on:click="$emit('submitParams')"
+                    type="button"
+                    :disabled="requiredMissing"
+                    class="btn btn-primary">
+                    Save changes
+                </button>
+                <button
+                    v-if="loadingInstance"
+                    v-on:click="$emit('submitInstance')"
+                    type="button"
+                    :disabled="requiredMissing"
+                    class="btn btn-primary">
+                    Add this sequence
+                </button>
+                <button
+                    v-on:click="$emit('closeModal')"
+                    type="button"
+                    class="btn btn-secondary"
+                    data-dismiss="modal">
+                    Close
+                </button>
             </div>
         </div>
     </div>
 </template>
 
 <script lang="ts">
     import Vue from 'vue'
-    import {ParamType} from '@/shared/ParamType'
+    import p5 from 'p5'
+    import {ParamInterface} from '@/shared/Paramable.ts'
 
     export default Vue.extend({
         name: 'SeqVizParamsModal',
-        methods: {},
+        methods: {
+            groupClass(indent: number): string {
+                let retval = 'form-group'
+                if (indent) {
+                    retval += ` indent${indent}`
+                }
+                return retval
+            },
+            setBigint(e: Event, p: {value: BigInt}) {
+                const target = e.target as HTMLInputElement
+                try {
```

Comment:
I don't understand this. Why is the catch block empty? This method will work even if `target` can't be converted to `BigInt`? What is `p` in this context? Why is `p.value` a substitute for `target`?

Reply:
There's nothing to prevent someone from typing, say, letters in a BigInt input area, in which case this handler has to do something -- so what it does is use the old value (what else could it do?) Hope I'm making sense?

Reply:
Sorry if I'm just out of the loop, but as far as the user experience, they should get an error if they type garbage, and not be allowed to proceed thinking that was ok.  How will that work if the code just uses the default value?

Reply:
Sorry if I wasn't clear. This handler responds on (virtually) every keystroke. So if the field is empty, and you type a 1, the parameter becomes 1. Then you type 0, it becomes 10. Then you type a 5, it becomes 105. Then you type an x say. 105x doesn't parse as a bigint, so the last value is used, and the field stays 105, which has the effect of making the x disappear after a beat. So effectively you are prevented from entering anything but a BigInt, which I think is generally considered a more effective UI than accepting ill formed stuff and then telling the user "sorry no good change that". Try it-- I think you will like the "feel"

Reply:
Oh that clears it up, thank you!  For bigints, any initial substring of a bigint is again a bigint, so this approach works fine.  Is there any situation where that's a problem?

Reply:
Well, there are custom handlers for a variety of types because of that sort of thing; this one is just for BigInt. We may need to add more handlers if we want to deal with other types in the future, but I am pretty sure I got all of the current ones.

Reply:
> So effectively you are prevented from entering anything but a BigInt, which I think is generally considered a more effective UI than accepting ill formed stuff and then telling the user "sorry no good change that". Try it-- I think you will like the "feel"

This isn't a hill I'm willing to die on, but I think an input should bark at you if you don't supply the correct input. Silently using the old value of 105 when the input is 105x seems very wrong to me. It should be the user's job to provide correct input. It should be the code's job to parse the input, not interpret the user's input. If the input is 105x, the user should see "Can't parse 105x.". That being said, I can't imagine this will be an issue all that often, so having aired my qualms, I'm happy to hold my nose and merge.

Reply:
What about having it show in red when the input is no good, so as soon as you type 105x, it shows it in red?  I haven't tried it yet, so this comment may be inapplicable somehow.  I'd like to pull it and play with it before we merge; I'll try to do that ASAP and maybe comment again here.

Reply:
> Silently using the old value of 105 when the input is 105x seems very wrong to me. It should be the user's job to provide correct input. It should be the code's job to parse the input, not interpret the user's input.

Again, that's not how it works. You type 105x, you see the 'x' briefly, and then it disappears because it's invalid. What you see is always what you get. There is no silent modification of what you type; there's constant validation so it only lets you type valid input. Have you tried it? What did you think of the user experience? From what I've read, earlier validation and an avoidance of the possibility of incorrect input is generally considered better UX.

Reply:
OK, Liam looked at this during our meeting and was comfortable with it, so I am re-resolving this conversation, although definitely I recommend @katestange also try it in practice _after_ I take care of the items below and _before_ this is merged.